### PR TITLE
Creative: Optimize some functions

### PR DESCRIPTION
**Pull request by motherfscker**
Before, changing of page, changing of (already visited) tab or clearing the filter triggered a lookup over all registered items and performed a table sorting inside the function `update_creative_inventory`.

Now, items are checked and sorted **strictly only when necessary** to minimize the overhead. This means :

- For all players : no lookup and table reconstruction when changing of tab, if one player already loaded that tab and if the filter is empty.
- No lookup and table reconstruction when changing of page.
- Clearing the filter directly applies the original item list by respective tab (common to all players).

The difference, performance-wise, won't probably be big when using LuaJIT and Minetest Game only. But it can be with lots of mods installed and regular Lua.

This PR comes also with some minor cleanups.

### CODE REVIEW

**Comment by sofar** [comment]
I object to having the following lines present in all of our future commit history:

```
From: motherfscker <yes@no.idk>
```

Shit like that gets minetest excluded from educational projects in schools.

If you're so afraid of showing your actual identity, please do not contribute to minetest.

Clearly this was a sham account created for the very purpose of making something like this slip in.

File: mods/creative/inventory.lua [hunk 1]
```diff
@@ -1,5 +1,7 @@
 creative = {}
-local player_inventory = {}
+local creative_mode = minetest.setting_getbool("creative_mode")
+local player_inventory, original_list = {}, {}
+local max, floor, ceil, sort = math.max, math.floor, math.ceil, table.sort
```
Comment (sapier) [inline mods/creative/inventory.lua]: this code ain't performance critical thus this replacement is not required.
If it's more readable is surely a matter of taste but I'd prefere math.max to the simplified variant.

File: mods/creative/inventory.lua [hunk 2]
```diff
@@ -172,9 +180,5 @@ creative.register_tab("craftitems", "Items", minetest.registered_craftitems)
 
 local old_homepage_name = sfinv.get_homepage_name
 function sfinv.get_homepage_name(player)
-	if minetest.setting_getbool("creative_mode") then
-		return "creative:all"
-	else
-		return old_homepage_name(player)
-	end
+	return creative_mode and "creative:all" or old_homepage_name(player)
```
Comment (sapier) [inline mods/creative/inventory.lua]: does this really provide same result?

File: mods/creative/inventory.lua [hunk 3]
```diff
@@ -11,11 +13,7 @@ function creative.init_creative_inventory(player)
 
 	minetest.create_detached_inventory("creative_" .. player_name, {
 		allow_move = function(inv, from_list, from_index, to_list, to_index, count, player2)
-			if not to_list == "main" then
-				return count
-			else
-				return 0
-			end
+			return to_list ~= "main" and count or 0
```
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: Please do not use this, as it makes code less readable.

File: mods/creative/inventory.lua [hunk 4]
```diff
@@ -79,17 +88,19 @@ function creative.register_tab(name, title, items)
 	sfinv.register_page("creative:" .. name, {
 		title = title,
 		is_in_nav = function(self, player, context)
-			return minetest.setting_getbool("creative_mode")
+			return creative_mode
```
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: What if the creative_mode setting is changed during play?

File: mods/creative/inventory.lua [hunk 5]
```diff
@@ -128,23 +139,20 @@ function creative.register_tab(name, title, items)
 			if fields.creative_clear then
 				inv.start_i = 0
 				inv.filter = ""
-				creative.update_creative_inventory(player_name, items)
```
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: Why is this removed?

File: mods/creative/inventory.lua [hunk 6]
```diff
@@ -128,23 +139,20 @@ function creative.register_tab(name, title, items)
...
 				sfinv.set_player_inventory_formspec(player, context)
 			elseif fields.creative_search or
 					fields.key_enter_field == "creative_filter" then
 				inv.start_i = 0
 				inv.filter = fields.creative_filter:lower()
-				creative.update_creative_inventory(player_name, items)
```
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: Same

File: mods/creative/inventory.lua (hunk 2 above)
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: @sapier it should provide the same result if creative_mode isn't changed during play. However, as it is less readable, I'd prefer this syntax would not be used for ternary conditionals (especially that it can be tricky, i.e. `a and b or c` is *not* the same as `a?b:c` if `b` evaluates to a false-y value).

**Comment by sapier** [comment]
I agree to sofar, if you don't want to say your name that's perfectly fine for me, but why so you use a that obviously offending name?

File: mods/creative/inventory.lua (hunk 2 above)
Comment (motherfscker) [inline mods/creative/inventory.lua]: Same result. Ternary conditional.

File: mods/creative/inventory.lua (hunk 4 above)
Comment (motherfscker) [inline mods/creative/inventory.lua]: Good point.

File: mods/creative/inventory.lua (hunk 5 above)
Comment (motherfscker) [inline mods/creative/inventory.lua]: It's actually a fix. `update_creative_inventory` is already called inside the `get` function provided by `sfinv`. So it was calling `update_creative_inventory` twice needlessly.

File: mods/creative/inventory.lua (hunk 6 above)
Comment (motherfscker) [inline mods/creative/inventory.lua]: See explanation above.

File: mods/creative/inventory.lua (hunk 2 above)
Comment (Ekdohibs) [inline mods/creative/inventory.lua]: Here the result is the same because `b` which is `"creative:all"` evaluates to a truth-y value; however, this is not always the case and leads this way to write ternary conditionals misleading. Moreover, this is only a style change which I would prefer you do not include (precisely for the reason that avoiding ternaries when possible, especially if it is a syntax not designed for ternaries).

**Comment by motherfscker** [comment]
Seeing the overreactions and the divergences of style, I encourage someone else to take this PR over. Sorry for the inconvenience.

**Comment by sofar** [comment]
please close this and lock conversation.

**Comment by Ekdohibs** [comment]
Please re-open this PR with another account with an username that is not offending if you want to get it merged. Closing this issue now.

**Comment by motherfscker** [comment]
If you really want to sacrifice — against your own interest (not really mine, I don't care) — a set of fixes and performance bonus for a ridiculous username, that is your own problem.

The message matters much more than the messenger. Except maybe for the superficial people paying too much attention to the appearances and not to the essential of things.

Do whatever you want with this PR now.

**Comment by octacian** [comment]
I'm sorry, but this is kinda ridiculous. You submitted a PR with a clearly offensive username, and the devs kindly asked you to change it and re-open the PR with another account/username. In case you haven't realized, Minetest _is_ used in several schools to teach kids programming, and it would not help if they discovered this commit in the history if they were to learn about Git.

Seriously, just change your username and provide a real email, it'll only take a minute.

**Comment by rubenwardy** [comment]
You don't need to provide a real email, a fake one is fine. You can use username@users.noreply.github.com
